core/iwasm/interpreter/wasm_loader.c

```diff
@@ -611,6 +626,16 @@ load_type_section(const uint8 *buf, const uint8 *buf_end, WASMModule *module,
             }
             type->param_cell_num = (uint16)param_cell_num;
             type->ret_cell_num = (uint16)ret_cell_num;
+
+            /* If there is already a same type created, use it instead */
+            for (j = 0; j < i; j++) {
+                if (wasm_type_equal(type, module->types[j])) {
```

Comment:
Had better check possible integer overflow before L633:
```C
if (module->types[j]->ref_count == UINT16_MAX) {
    set_error_buf(error_buf, error_buf_size,
                  "wasm type's ref count too large");
    return false;
}
```
When module->types[j]->ref_count reaches UINT16_MAX, `module->types[j]->ref_count++` will be 0.